MPI plugin docs (#1095)
NVIDIA/cuda-quantum: docs/sphinx/install.rst

a custom CUDA Quantum MPI plugin targeting their local MPI installation without recompiling or reinstalling the stack.

CUDA Quantum MPI API for C++ and Python will then be delegated 
to the plugin implementation selected at runtime.  

The following table summarizes mechanisms whereby a CUDA Quantum MPI plugin can be picked up.

.. tab:: Builtin

  **Use case**: build from source, Docker containers

  When a local MPI installation is detected at build time, CUDA Quantum will build an MPI plugin targeting this implementation when build from source.
  Similarly, the :ref:`CUDA Quantum Docker image <install-docker-image>` is shipped with a builtin MPI plugin based on its optimized OpenMPI installation inside the image. 
  No action is required to use this builtin plugin. 

  .. note::

    Manually-activated MPI plugin takes precedence over builtin plugin. 
    For instance, the user may choose to reinstall a different MPI library inside the Docker container and 
    activate a new CUDA Quantum MPI plugin. The newly-activated plugin will always be used even though a builtin one 
    was shipped with the Docker image. 

.. tab:: Manual Activation

>   **Use case**: data centers with custom MPI installation, CUDA Quantum binary distributions without MPI support (including Python wheels).

Reviewer: Does it include also users using the docker image but have a local MPI implementation that is not open MPI?

Author: Yes, it does. The manually-activated MPI plugin (targeting a local MPI implementation) will override the built-in one in the docker image (based on OpenMPI).

>   **Use case**: data centers with custom MPI installation, CUDA Quantum binary distributions without MPI support (including Python wheels).

Reviewer: Curious to why we don't have MPI support in binary distributions and Python wheels

Author: 'CUDA Quantum binary distributions' here refers to general distributions that may or may not have MPI bundled. Python wheel distribution is one example where we don't currently have MPI in the distributed binary. 
I think the main reason is that the default Python `pip` package manager doesn't support non-Python packages, like MPI, as a package dependency.

Reviewer 2: Re Python wheels: wheels have a pretty strict size limit, and it is hence expected that wheels do take dependencies on other wheels rather than include the same content. If you look at the [mpi4py page on PyPI](https://pypi.org/project/mpi4py/) you find that the wheel itself only contains the bindings, but expects an existing MPI installation on the system - like I already mentioned regarding CUDA. It hence makes little sense for us to declare that wheel as a dependency. A better way to get a fully functional MPI installation usable from Python is to follow the instructions we give, which install a *CUDA-aware!* MPI implementation along with the python bindings using conda.

  **Instructions:**

  (1) Locate the CUDA Quantum `distributed_interfaces` sub-directory at the top-level of the CUDA Quantum installation directory.

>   (2) Execute the `activate_custom_mpi.sh` bash script.

Reviewer: What is there are more than one local MPI implementation available? (is that possible?)
Is there a way to specify which one we want to use?

    In particular, `${MPI_PATH}/include` is expected to contain the `mpi.h` header and 
    `${MPI_PATH}/lib64` or `${MPI_PATH}/lib` is expected to contain `libmpi.so`.

  .. note::

    After the initial activation, if performed from the CUDA Quantum install directory,the activated will always be used to handle CUDA Quantum MPI API.

  .. note::

    Executing the activation script from the CUDA Quantum installation directory requires *write* permissions to that directory.
    If that is not feasible, the user might copy the `distributed_interfaces` sub-directory to a local location and execute the 
    activation script from there.

    In this scenario, since the activated plugin (`libcudaq_distributed_interface_mpi.so`) is outside the CUDA Quantum installation,
    CUDA Quantum requires users to set an environment variable `$CUDAQ_MPI_COMM_LIB` containing the path to that shared library.
    This is done automatically when executing that activation script. Users may wish to persist that environment variable 
    between bash sessions, e.g., via the `.bashrc` file.

.. tab:: `mpi4py <https://mpi4py.readthedocs.io/>`__ Wrapper

  **Use case**: Python users

>   Although we recommend manual MPI activation for optimized performance and compatibility, users may wish

Reviewer: What is the performance impact of using mpi4py over manual activation? 
Maybe write it differently by removing this part and at the end specifying that it might have a perf hit and why.

Author: Good point. I'll rewrite this part.

we recommend installing the complete CUDA toolkit (`cuda-toolkit-11-8`) that also 
includes the `nvcc` compiler.

.. _distributed-computing-with-mpi:

Distributed Computing with MPI
------------------------------------

CUDA Quantum supports the Message Passing Interface (MPI) parallelism via a runtime configurable plugin interface.
Specifically, while a builtin plugin is shipped with the CUDA Quantum Docker image, users can also activate 
a custom CUDA Quantum MPI plugin targeting their local MPI installation without recompiling or reinstalling the stack.

CUDA Quantum MPI API for C++ and Python will then be delegated 
to the plugin implementation selected at runtime.  

The following table summarizes mechanisms whereby a CUDA Quantum MPI plugin can be picked up.

.. tab:: Built-in

  **Use case**: build from source, Docker containers

>   When a local MPI installation is detected at build time, CUDA Quantum will build an MPI plugin targeting this implementation when build from source.

Reviewer 2: ```suggestion
  When a local MPI installation is detected at build time, CUDA Quantum will build an MPI plugin targeting this implementation.
```

>   Similarly, the :ref:`CUDA Quantum Docker image <install-docker-image>` is shipped with a builtin MPI plugin based on its optimized OpenMPI installation inside the image. 

Reviewer 2: ```suggestion
  Similarly, the :ref:`CUDA Quantum Docker image <install-docker-image>` is shipped with a built-in MPI plugin based on an optimized OpenMPI installation included in the image. 
```

  No action is required to use this built-in plugin. 

  .. note::

>     Manually-activated MPI plugin takes precedence over builtin plugin. 

Reviewer 2: ```suggestion
    A manually activated MPI plugin takes precedence over a built-in plugin. 
```

    For instance, the user may choose to reinstall a different MPI library inside the Docker container and 
    activate a new CUDA Quantum MPI plugin. The newly-activated plugin will always be used even though a built-in one 
    was shipped with the Docker image. 

.. tab:: Manual Activation

  **Use case**: 
  
  (1) Using the CUDA Quantum Docker image on systems that have a different MPI implementation.
  For instance, HPC data centers often have a vendor-optimized MPI library pre-installed on
  their system, which need to be swapped with the MPI library in the Docker image at runtime.
  With manual activation, a custom CUDA Quantum MPI plugin can be built against the system MPI library, 
  guaranteeing runtime compatibility when MPI injection into the container occurs. 

  (2) Post-deployment MPI activation for CUDA Quantum binaries that do not have MPI support.
  Depending on the distribution channel, CUDA Quantum binaries may not come with any pre-built
  MPI plugin. For example, CUDA Quantum Python wheels from PyPI and :ref:`pre-built binaries <install-prebuilt-binaries>` 

>   do not have a built-in MPI plugin. CUDA Quantum MPI capabilities can be activated post-installation

Reviewer 2: ```suggestion
  do not include a built-in MPI plugin. CUDA Quantum MPI capabilities can be activated post-installation
```

  against the MPI library available on the local system.

  **Instructions:**

  (1) Locate the CUDA Quantum `distributed_interfaces` sub-directory at the top-level of the CUDA Quantum installation directory.

  (2) Execute the `activate_custom_mpi.sh` bash script.

  .. code-block:: console

    source $CUDAQ_INSTALL_DIR/distributed_interfaces/activate_custom_mpi.sh

  This will build and install a custom MPI plugin targeting the currently-available MPI implementation.
  
  .. note::

    This activation script requires the `MPI_PATH` environment variable being set to locate the MPI installation.
    In particular, `${MPI_PATH}/include` is expected to contain the `mpi.h` header and 
    `${MPI_PATH}/lib64` or `${MPI_PATH}/lib` is expected to contain `libmpi.so`.

  .. note::

    After the initial activation, if performed from the CUDA Quantum install directory,the activated will always be used to handle CUDA Quantum MPI API.

  .. note::

    Executing the activation script from the CUDA Quantum installation directory requires *write* permissions to that directory.

>     If that is not feasible, the user might copy the `distributed_interfaces` sub-directory to a local location and execute the 

Reviewer 2: ```suggestion
    If you do not have the necessary permissions, copy the `distributed_interfaces` sub-directory to a local location and execute the 
```

    activation script from there.

    In this scenario, since the activated plugin (`libcudaq_distributed_interface_mpi.so`) is outside the CUDA Quantum installation,
    CUDA Quantum requires users to set an environment variable `$CUDAQ_MPI_COMM_LIB` containing the path to that shared library.
    This is done automatically when executing that activation script. Users may wish to persist that environment variable 
    between bash sessions, e.g., via the `.bashrc` file.

.. tab:: mpi4py Wrapper

  **Use case**: Python users 

>   If manual activation of CUDA Quantum MPI plugin is not feasible, users may attain MPI support via the 

Reviewer 2: ```suggestion
  If you have not manually activated a CUDA Quantum MPI plugin can attain MPI support via the 
```

  the `mpi4py <https://mpi4py.readthedocs.io/>` Python package. On a system with no builtin nor manually-activated plugin, 

>   CUDA Quantum will try to locate the `mpi4py` package.

Reviewer 2: ```suggestion
  CUDA Quantum will automatically try to locate the `mpi4py` package.
```
